### PR TITLE
Merge renamed files

### LINKED ISSUE #14: dealing with renamed files
Add function to set value of the column `changed_file` to the current name of the files, so that grouping by `changed_file` groups all file-specific commit information for a (potentially multiple times) renamed file into one group.

### PR DESCRIPTION
**Pull request by lorenzwalthert**
Aims to close #14. The problem is described in the help file for the newly exported function `set_changed_file_to_latest_name()`:
> parse_log_detailed() parses a git log. However, changed file names are not corrected retroactively. Hence, if one want's to process file-specific commit data for one file accross different names for that file, the column changed_files has to be unified. This is done with this function by updating the file names recursively and setting them to the latest name of them.

### CODE REVIEW

**Comment by lorenzwalthert** [comment]
Todo: add tests

File: R/renamed-files.R [hunk 1]
```diff
@@ -0,0 +1,152 @@
+#' Set the column  `changed_files` of a `log` to the latest name of the file
+#'
+#' [parse_log_detailed()] parses a git log. However, changed file names
+#' are not corrected retroactively. Hence, if one want's to process file-specific
```
Comment (jonmcalder) [inline R/renamed-files.R]: `want's` shouldn't have an apostrophe

File: R/renamed-files.R [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  contains_renaming <- log[is_renaming, ]
+  contains_no_renaming <- log[!is_renaming,]
+  transition_model <- parse_reassignment(
+    contains_renaming$changed_file, which(is_renaming)
+  )
+  update_changed_file_sequentially(log, transition_model)
+}
+
+
+#' Detetct whether a `changed_file`-entry is a renaming
```
Comment (jonmcalder) [inline R/renamed-files.R]: Typo in "Detect"

File: R/renamed-files.R [hunk 3]
```diff
@@ -0,0 +1,152 @@
+#' Set the column  `changed_files` of a `log` to the latest name of the file
+#'
+#' [parse_log_detailed()] parses a git log. However, changed file names
+#' are not corrected retroactively. Hence, if one want's to process file-specific
+#' commit data for one file accross different names for that file, the column
```
Comment (jonmcalder) [inline R/renamed-files.R]: Typo in "across"

**Comment by lorenzwalthert** [comment]
Thanks.

**Comment by lorenzwalthert** [comment]
Did you @jonmcalder get a chance to look at the non-documentative code?

**Comment by lorenzwalthert** [comment]
Ok, tests are added.

File: R/renamed-files.R [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+#'
+#' @param changed_file The column `changed_file` from a parsed unnested
+#'   detailed log.
+#' @param reassignment_pattern The regular expression pattern that corresponds
+#'   to renaming.
+#' @importFrom stringr str_locate
+#' @examples
+#' gitsum:::is_name_change("R/{gitsum.R => gitsum-package.R}")
+is_name_change <- function(changed_file,
+                           reassignment_pattern = "\\{.+\\s\\=\\>.+\\}$") {
```
Comment (jonmcalder) [inline R/renamed-files.R]: I noticed that this pattern might miss some changed files due to truncation of the `changed_file` column in the log.

e.g. see [man/ensure_gitusm_repo.Rd → man/ensure_gitsum_repo.Rd](https://github.com/lorenzwalthert/gitsum/commit/bc424b294eb239a770d683a5dba5f0103668ded5#diff-2f53785f2310dc1682ed4bce1e134c45)

To verify that this get's missed run the below code from within your gitsum repo:

```
unnest_log <- parse_log_detailed(".") %>% tidyr::unnest()
unnest_log[is_name_change(unnest_log$changed_file),"changed_file"]
# A tibble: 10 x 1
...
unnest_log[grepl("=>", unnest_log$changed_file),"changed_file"]
# A tibble: 11 x 1
                                         changed_file
                                                <chr>
 1              man/{get_log.Rd => get_log_simple.Rd}
 2             man/{parse_log.Rd => parse_log_one.Rd}
 3    man/{get_pattern.Rd => get_pattern_multiple.Rd}
 4      man/{get_log_regex.Rd => git_log_detailed.Rd}
 5               man/{rmd_simple.Rd => git_report.Rd}
 6                   R/{gitsum.R => gitsum-package.R}
 7 man/{git_log_detailed.Rd => parse_log_detailed.Rd}
 8     man/{git_log_simple.Rd => parse_log_simple.Rd}
 9 ...ensure_gitusm_repo.Rd => ensure_gitsum_repo.Rd}
10                       man/{log.Rd => gitsumlog.Rd}
11               man/{git_report.Rd => report_git.Rd}
```
Do you think it's necessary to check for `{` and `}`? Maybe this regex could/should be relaxed slightly?

File: R/renamed-files.R [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+#' Parse a raw reassignment into a list
+#'
+#' Takes a vector of raw reassignments and turns them into a list of elements.
+#' Every element contains the old path, the new path and the index of the
+#' reassignment.
+#' @param raw_reassignment Character vector with raw re-assignments.
+#' @param reassignment_index Integer vector indicating the position of the
+#'   re-assignment in the original unnested detailed log.
+#' @examples
+#' gitsum:::parse_reassignment(rep("R/{a => b}", 2), c(1, 2))
```
Comment (jonmcalder) [inline R/renamed-files.R]: What about renaming which results in a file moving between different sub-directories?

**Review by jonmcalder — CHANGES_REQUESTED** [review]
As things stand I've only really reviewed parts of this code, but I think maybe it's worth waiting for you to address my comments so far?

File: R/renamed-files.R (hunk 4 above)
Comment (lorenzwalthert) [reply R/renamed-files.R]: Thanks for pointing out to that. I have just pushed a quick fix in #19. It essentially sets the environment variable `$COLUMNS` to a high value before writing the log to a file and hence should avoid truncation of the output. Not sure whether that's the way to go. Also, we should probably check that there is no truncated `changed_file` somehow.

Comment (lorenzwalthert) [reply R/renamed-files.R]: I just changed the name of a root directory file and the file name with the reassignment  it does not have curly braces, so you are right, we indeed should not require curly braces.

File: R/renamed-files.R (hunk 5 above)
Comment (lorenzwalthert) [reply R/renamed-files.R]: I think the prefix before `{` will always be the "least common root", but there won't be braces if the file is in the root directory (https://github.com/lorenzwalthert/gitsum/pull/15#discussion_r160056711)

**Comment by lorenzwalthert** [comment]
Sure.

**Comment by lorenzwalthert** [comment]
Abandoned in favour of #29.

**Pull request closed without merging**